### DISCUSSION TITLE
[Discussion] Composition API Re-write

### DISCUSSION THREAD

**Discussion by DerrikMilligan** [antoniandre/wave-ui#130] [Ideas]
Hey @antoniandre you mentioned in the typescript issue that "_In a few months, I will stop supporting the legacy branch and we can rewrite everything in composition API in the main branch._"

I wanted to open this issue as a discussion about it. Are you looking to take on this task single-handedly? Or would you like help in the effort? If you're open to help, I'd be more than willing to help and there may be others who are as well. We'd probably just need some sort of a game plan on how we want things to work/look and then we can start converting components.

Also I wanted to open a dialog about a couple things that could potentially be changed/improved as we rewrite it. I LOVE the framework and how it is simple and works to not overstep it's bounds with CSS which is one of the main reasons I started using it. There is always room for improvement however. And a re-write is a good time to discuss or think about any larger changes we would want to consider since it'll likely not have to worry about backwards compatibility.

The one thing that has been one my mind as something that would be a potential improvement would be to use [Floating UI](https://floating-ui.com/) (The successor of [popper.js](https://popper.js.org/)) to handle the positioning and logic for all the `detachable` type components. They even have it ready to integrate with [Vue](https://floating-ui.com/docs/vue) It's battle tested and has all the options anyone could ever want. The one pain point I've experienced with a few of the detachable components is then going outside the view-port.

I've seen that in the `computeDetachableCoords` method in `src/wave-ui/mixins/detachable.js` there is a code block commented out as `Step 3` that outlines needing to prevent something from displaying outside the view-port. This is some tricky math and I wonder if outsourcing it to a very tiny library that is REALLY good at handling those things.

That is the only real larger change I've seen a potential need for. Everything else has been small and easy enough to make a pull request for to fix up.

**Comment by antoniandre** [comment]
Hi @DerrikMilligan, thanks for opening this discussion.
Indeed we will be able to rewrite the codebase with composition API when I stop maintaining the `legacy` branch.
No definitive date right now but probably talking about a few months away.

I am open to help, as long as we keep the same codebase style and spirit and that we stick to the same functionality, or discuss changes first. Wave UI has an underlying philosophy that any commit should respect: lean code, not too edge case, and CSS-ease first. :) (@DerrikMilligan has already made lots of contributions and always respecting these rules)

Regarding your idea about using a dependency, it's not going to be the direction we will take. Wave UI has and will remain dependency free. 
I also have a couple of other libs that could be a great addition to Wave UI (like Splitpanes or Vueperslides), but Wave UI's original purpose is to scaffold UIs fast and simply, and not to grow in size by adding tons of components supporting themselves tons of options - which implies less external CSS flexibility when overriding.
Adding dependencies, will grow the size and could possibly bring troubles by not owning the external code when something needs to be modified/improved. It should also be a user's decision to add external libs IMO.

That being happily said, you can take a stab at making the commented code work in the detachable mixin, with or without insipration/pieces of codes from elsewhere as long as it's not too much overhead in Wave UI! :)

As always, thank you for your help! 🤗
